daemon/containerd/image_builder.go

```diff
@@ -192,23 +191,59 @@ func newROLayerForImage(ctx context.Context, imgDesc *ocispec.Descriptor, i *Ima
 		return nil, err
 	}
 
+	// TODO(vvoland): Check if image is unpacked, and unpack it if it's not.
 	imageSnapshotID := identity.ChainID(diffIDs).String()
 
+	snapshotter := i.StorageDriver()
+	lease, err := leaseSnapshot(ctx, i.client.LeasesService(), snapshotter, imageSnapshotID)
+	if err != nil {
+		return nil, errdefs.System(fmt.Errorf("failed to lease image snapshot %s: %w", imageSnapshotID, err))
+	}
+
 	return &rolayer{
 		key:                imageSnapshotID,
 		c:                  i.client,
-		snapshotter:        i.snapshotter,
+		snapshotter:        snapshotter,
 		diffID:             "", // Image RO layer doesn't have a diff.
 		contentStoreDigest: "",
+		lease:              &lease,
 	}, nil
 }
 
+func leaseSnapshot(ctx context.Context, lm leases.Manager, snapshotter string, snapshotID string) (leases.Lease, error) {
+	if snapshotID == "" {
+		return leases.Lease{}, nil
```

Comment:
I wonder if this is correct now, because we never return `nil`, so the idea of `lease` being `nil` ("there's no lease") may not apply now?

Should we change the signature of this function to return a pointer?

```go
func leaseSnapshot(ctx context.Context, lm leases.Manager, snapshotter string, snapshotID string) (*leases.Lease, error) {
	if snapshotID == "" {
		return nil, nil
```

Reply:
Ah right, good catch!